pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -581,7 +581,9 @@ void ConsumerImpl::doAcknowledge(const BatchMessageId& messageId, proto::Command
 
 void ConsumerImpl::disconnectConsumer() {
     LOG_DEBUG("Broker notification of Closed consumer: " << consumerId_);
+    Lock lock(mutex_);
```

Comment:
handleCreateConsumer "connection_ = cnx" and disconnectConsumer "connection_.reset()" can happen at the same time on different threads, hence the lock